### ISSUE TITLE
[TTS-JETS] about CTCLoss?

### ISSUE THREAD

**Issue by BridgetteSong** [espnet/espnet#5179]
1.  document of [torch.nn.CTCLoss](https://pytorch.org/docs/stable/generated/torch.nn.CTCLoss.html#torch.nn.CTCLoss) says: the input is a **logarithmized probabilities**, obtained with `torch.nn.functional.log_softmax()`
2. when added **blank_value** and **beta-binomial prior**, the **log_p_attn_pd** is not a log-probabilities,  it shoud be:

- `cur_log_p_attn_pd = torch.nn.functional.log_softmax(cur_log_p_attn_pd, dim=-1)`

-  `loss += F.ctc_loss(log_probs=cur_log_p_attn_pd,targets=target_seq,
                input_lengths=olens[bidx : bidx + 1],
                target_lengths=ilens[bidx : bidx + 1],
                zero_infinity=True)`

https://github.com/espnet/espnet/blob/d55008a8b2e90ae6c92a4232bff530521ca19c51/espnet2/gan_tts/jets/loss.py#LL152C1-L170C14

**Comment by sw005320** [comment]
@kan-bayashi and @imdanboy, can you confirm it?

As @BridgetteSong mentions, `log_p_attn_pd` is not a log-probability, but in my understanding, whether the log-probability or not is not a matter for computing the CTC loss.
But the log-probability is a good way to normalize the logit to avoid the numerical issue.
Since all values are already based on the log probability (normalized), their addition is still somewhat normalized (but not the log probability), and it would be no problem.

**Comment by imdanboy** [comment]
Hi, `log_p_attn` is a log-probability and is an output of `AlignmentModule`
https://github.com/espnet/espnet/blob/master/espnet2/gan_tts/jets/alignments.py#L58
As above, it is already an output of `log_softmax` function.

**Comment by BridgetteSong** [comment]
@imdanboy 
Yes, `log_p_attn` is  a log-probability output, but when before you feed it in to `CTCLoss()`, it becomes `log_p_attn + bb_prior`, and then pad a `blank` in the 0th, so `log_p_attn_pd` is not a log-probability output, because `sum(exp(log_p_attn_pd)) != 1`. 

In my experiment, when  the input of `CTCLoss()` is not a log-probability, loss will become very big(), but when it is a log-probability, it can be 1.5 when your `ctc_loss_weight ` is 2.0.

[torch.nn.CTCLoss()](https://pytorch.org/docs/stable/generated/torch.nn.CTCLoss.html#torch.nn.CTCLoss) needs the input is a log-probability, see [Nvidia-FastPitch](https://github.com/NVIDIA/DeepLearningExamples/blob/9becdf8d53aac528cc5e214e27314f2572971d48/PyTorch/SpeechSynthesis/FastPitch/fastpitch/attn_loss_function.py#LL51C1-L51C54)

BTW, it is very difficult to train model if feed `text_decoder_outputs `into `generator`, and `MelLoss `is very big(about 35.0 when `c_mel `is 45.0). how to solve this?

**Comment by imdanboy** [comment]
@BridgetteSong 
Oh I haven't thought about it, you are right. Thanks your kind explanation!

Regarding above reply by @sw005320, I am not sure there are strict requirement (`sum(exp(log_probs)) == 1`) in `CTCLoss` for correct value.
I just guess current value of CTC loss is still valid in my understanding of CTC algorithm,
so the current value of CTC loss is same as value of CTC loss on log-probability with more high `ctc_loss_weight`?

In my case, JETS got reasonable results on LJSpeech, KSS recipe with big mel loss value compared to other loss terms. (JETS feed `text_decoder_outputs` into `generator` right?)

**Comment by BridgetteSong** [comment]
@imdanboy In the test experiment just now, I observed the `CTCLoss `value by using an `unnormalized log-probability`(same as yours), it can decrease into about **5.0 in the first 1000 steps**, but when a `normalized log-probability` was feeded in to `CTCLoss`, the max loss value is about **3.0 in the first 200 steps** and fast decreased to **1.0 in the 1000 steps**.

So I guess when an `unnormalized log-probability` is feeded into `CTCLoss`, the **CTC algorithm** fails to train, but you can still get a good result, because The learning durations module degenerates into a easy **MAS**(same as **GlowTTS** or **VITS**).

In a **single-speaker high quality dataset**, I can get a relatively good result(with **MelLoss** about **25.0**). But when I use a **multi-speakers low quality dataset**, the model is difficult to converge with a **MelLoss** about **35.0**.

**Comment by imdanboy** [comment]
@BridgetteSong 
Thank you for sharing your results.
The current implementation might be somewhat wrong, it would be better to follow your suggestions.
I will check it out too!
